File: Models/QuestionDataManager.swift
import Foundation

/// Gestionnaire optimisé pour les données de questions
/// Utilise les String Catalogs (.xcstrings) avec NSLocalizedString
class QuestionDataManager: ObservableObject {
    static let shared = QuestionDataManager()
    
    @Published var isLoading = false
    private var questionsCache: [String: [Question]] = [:]
    
    private init() {}
    
    // MARK: - Question Loading
    
    /// Charge les questions pour une catégorie spécifique
    func loadQuestions(for categoryId: String, language: String = QuestionDataManager.getCurrentLanguage()) -> [Question] {
        let cacheKey = "\(categoryId)_\(language)"
        
        // Vérifier le cache d'abord
        if let cachedQuestions = questionsCache[cacheKey] {
            return cachedQuestions
        }
        
        // Charger depuis les String Catalogs
        let questions = loadQuestionsFromStringCatalogs(categoryId: categoryId)
        questionsCache[cacheKey] = questions
        
        return questions
    }
    
    /// Récupère une chaîne localisée avec un fallback en anglais si introuvable dans la langue courante
    private func localizedString(_ key: String, tableName: String) -> String {
        // Première tentative avec la langue courante
        let primary = NSLocalizedString(key, tableName: tableName, bundle: .main, value: key, comment: "")
        if primary != key {
            return primary
        }
        // Fallback : essayer dans le bundle Anglais si disponible
        if let enPath = Bundle.main.path(forResource: "en", ofType: "lproj"),
           let enBundle = Bundle(path: enPath) {
            let secondary = NSLocalizedString(key, tableName: tableName, bundle: enBundle, value: key, comment: "")
            return secondary
        }
        return key // dernier recours : retourner la clé
    }

    /// Charge les questions depuis les String Catalogs (.xcstrings)
    private func loadQuestionsFromStringCatalogs(categoryId: String) -> [Question] {
        // Mapping des IDs de catégorie vers les préfixes des clés
        //  et vers le nom du String Catalog (tableName)
        let mapping: [String: (prefix: String, table: String, start: Int)] = [
            "en-couple": ("ec_", "EnCouple", 2),
            "les-plus-hots": ("lph_", "LesPlus Hots", 2),
            "pour-rire-a-deux": ("prad_", "PourRire", 2),
            "questions-profondes": ("qp_", "QuestionsProfondes", 2),
            "a-distance": ("ad_", "ADistance", 2),
            "tu-preferes": ("tp_", "TuPreferes", 2),
            "mieux-ensemble": ("me_", "MieuxEnsemble", 2),
            "pour-un-date": ("pud_", "PourUnDate", 2)
        ]
        
        guard let info = mapping[categoryId] else {
            print("⚠️ QuestionDataManager: Catégorie inconnue: \(categoryId)")
            return []
        }
        
        let keyPrefix = info.prefix
        let tableName = info.table
        let startIndex = info.start
        
        // Générer les questions en utilisant la fonction helper avec fallback
        var questions: [Question] = []
        for i in startIndex...300 {
            let key = "\(keyPrefix)\(i)"
            let localizedText = localizedString(key, tableName: tableName)
            if localizedText != key && !localizedText.isEmpty {
                questions.append(Question(id: key, text: localizedText, category: categoryId))
            } else if i > startIndex + 10 && questions.isEmpty {
                break
            }
        }
        print("✅ QuestionDataManager: \(questions.count) questions chargées pour \(categoryId)")
        return questions
    }
    
    // MARK: - Utility Methods
    
    /// Obtient la langue actuelle de l'appareil
    private static func getCurrentLanguage() -> String {
        return Locale.preferredLanguages.first?.prefix(2).lowercased() == "fr" ? "fr" : "en"
    }
    
    /// Précharge toutes les catégories essentielles
    func preloadEssentialCategories() {
        let essentialCategories = ["en-couple"] // Seulement la catégorie gratuite
        let currentLanguage = QuestionDataManager.getCurrentLanguage()
        
        Task {
            isLoading = true
            
            for categoryId in essentialCategories {
                _ = loadQuestions(for: categoryId, language: currentLanguage)
            }
            
            DispatchQueue.main.async {
                self.isLoading = false
                print("QuestionDataManager: Catégories essentielles préchargées")
            }
        }
    }
    
    /// Vide le cache (utile pour les changements de langue)
    func clearCache() {
        questionsCache.removeAll()
    }
}

// MARK: - Data Models

struct QuestionData: Codable {
    let category: String
    let questions: [QuestionJSON]
}

struct QuestionJSON: Codable {
    let id: String
    let text: String
} 

File: Models/QuestionDataManager.kt
// QuestionDataManager.kt
package com.yourapp.data

import android.content.Context
import android.content.res.Configuration
import android.os.Build
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import java.util.Locale

/**
 * Gestionnaire optimisé pour les données de questions (Android).
 * Remplace l'usage de .xcstrings par strings.xml (resources Android).
 *
 * - Localisation : context.getString(R.string.key)
 * - Résolution dynamique : resources.getIdentifier("ec_2", "string", packageName)
 * - Fallback : si la ressource n'existe pas dans la langue courante, on tente l'anglais.
 */
object QuestionDataManager {

    // État de chargement (Compose-friendly)
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> get() = _isLoading

    // Cache: clé = "${categoryId}_${language}"
    private val questionsCache: MutableMap<String, List<Question>> = HashMap()

    // Mapping catégorie → (préfixe des clés, index de départ)
    private val CATEGORY_MAPPING: Map<String, CategoryInfo> = mapOf(
        "en-couple"           to CategoryInfo(prefix = "ec_",  startIndex = 2),
        "les-plus-hots"       to CategoryInfo(prefix = "lph_", startIndex = 2),
        "pour-rire-a-deux"    to CategoryInfo(prefix = "prad_",startIndex = 2),
        "questions-profondes" to CategoryInfo(prefix = "qp_",  startIndex = 2),
        "a-distance"          to CategoryInfo(prefix = "ad_",  startIndex = 2),
        "tu-preferes"         to CategoryInfo(prefix = "tp_",  startIndex = 2),
        "mieux-ensemble"      to CategoryInfo(prefix = "me_",  startIndex = 2),
        "pour-un-date"        to CategoryInfo(prefix = "pud_", startIndex = 2),
    )

    data class CategoryInfo(
        val prefix: String,
        val startIndex: Int
    )

    // -------- Public API --------

    /**
     * Charge les questions pour une catégorie donnée.
     * Le cache est indexé par catégorie + langue courante (fr/en).
     */
    fun loadQuestions(
        context: Context,
        categoryId: String,
        language: String = getCurrentLanguage(context)
    ): List<Question> {
        val cacheKey = "${categoryId}_$language"
        questionsCache[cacheKey]?.let { return it }

        val questions = loadQuestionsFromResources(context, categoryId)
        questionsCache[cacheKey] = questions
        return questions
    }

    /**
     * Précharge les catégories essentielles (gratuit).
     * Appelle-le depuis un ViewModel (scope) si tu veux lier à l’UI.
     */
    suspend fun preloadEssentialCategories(context: Context) {
        _isLoading.emit(true)
        try {
            val essential = listOf("en-couple")
            val lang = getCurrentLanguage(context)
            essential.forEach { loadQuestions(context, it, lang) }
        } finally {
            _isLoading.emit(false)
        }
    }

    /**
     * Vide le cache (utile après un changement de langue manuel).
     */
    fun clearCache() {
        questionsCache.clear()
    }

    // -------- Internal --------

    private fun loadQuestionsFromResources(
        context: Context,
        categoryId: String
    ): List<Question> {
        val info = CATEGORY_MAPPING[categoryId]
        if (info == null) {
            println("⚠️ QuestionDataManager: Catégorie inconnue: $categoryId")
            return emptyList()
        }

        val keyPrefix = info.prefix
        val startIndex = info.startIndex

        val found = mutableListOf<Question>()
        // On scanne jusqu’à 300 par cohérence avec ta version Swift.
        for (i in startIndex..300) {
            val key = "${keyPrefix}${i}"
            val text = localizedString(context, key)
            if (!text.isNullOrBlank()) {
                found += Question(id = key, text = text, category = categoryId)
            } else if (i > startIndex + 10 && found.isEmpty()) {
                // Même heuristique que ta version Swift : si rien trouvé
                // après un petit offset, on s’arrête tôt.
                break
            }
        }

        println("✅ QuestionDataManager: ${found.size} questions chargées pour $categoryId")
        return found
    }

    /**
     * Essaie de lire la ressource string nommée [key] dans la langue courante,
     * puis retombe sur l’anglais si indisponible.
     *
     * Équivalent Android de:
     *   challengeKey.localized(tableName: ...)  →  context.getString(R.string.challengeKey)
     * Ici, comme la clé est construite dynamiquement (ex: "ec_12"),
     * on utilise getIdentifier pour retrouver l'ID.
     */
    private fun localizedString(
        context: Context,
        key: String,
        englishFallback: Boolean = true
    ): String? {
        val resId = context.resources.getIdentifier(key, "string", context.packageName)
        if (resId == 0) return null // la clé n’existe pas du tout

        // 1) Tentative dans la langue courante
        val primary = safeGetString(context, resId)
        if (!primary.isNullOrBlank()) return primary

        // 2) Fallback anglais explicite (si souhaité)
        return if (englishFallback) getStringInLocale(context, resId, Locale.ENGLISH) else null
    }

    private fun safeGetString(context: Context, resId: Int): String? =
        try { context.getString(resId) } catch (_: Throwable) { null }

    private fun getStringInLocale(context: Context, resId: Int, locale: Locale): String? {
        return try {
            val conf = Configuration(context.resources.configuration)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                conf.setLocale(locale)
            } else {
                @Suppress("DEPRECATION")
                conf.locale = locale
            }
            val localizedCtx = context.createConfigurationContext(conf)
            localizedCtx.getString(resId)
        } catch (_: Throwable) {
            null
        }
    }

    /**
     * Détermine une étiquette simple de langue pour le cache ("fr" ou "en").
     * Android gère déjà automatiquement la sélection de ressources locales.
     * Ici c’est seulement pour éviter de mélanger le cache quand l’UI change de locale.
     */
    private fun getCurrentLanguage(context: Context): String {
        val lang = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            context.resources.configuration.locales.get(0)?.language
        } else {
            @Suppress("DEPRECATION")
            context.resources.configuration.locale?.language
        } ?: Locale.getDefault().language

        return if (lang.lowercase(Locale.ROOT).startsWith("fr")) "fr" else "en"
    }
}

// -------- Data Models --------

data class Question(
    val id: String,
    val text: String,
    val category: String
)

data class QuestionData(
    val category: String,
    val questions: List<QuestionJSON>
)

data class QuestionJSON(
    val id: String,
    val text: String
)
